sirtextstyle.c

```diff
@@ -117,187 +131,94 @@ bool _sir_resettextstyles(void) {
 
     bool all_ok = true;
     for (size_t n = 0; n < SIR_NUMLEVELS; n++) {
-        map[n].style = _sir_getdefstyle(map[n].level);
-        all_ok &= _sir_formatstyle(map[n].style, map[n].str, SIR_MAXSTYLE);
+        memcpy(&map[n].style, _sir_getdefstyle(map[n].level), sizeof(sir_textstyle));
+        all_ok &= _sir_formatstyle(mode, &map[n].style, map[n].str);
     }
 
     _sir_unlocksection(SIRMI_TEXTSTYLE);
     return all_ok;
 }
 
-uint16_t _sir_getprivstyle(sir_textstyle style) {
-    static const size_t idx_attr_start = 0;
-    static const size_t idx_attr_end   = 2;
-
-    static const size_t idx_fg_start = 3;
-    static const size_t idx_fg_end   = 19;
-
-    static const size_t idx_bg_start = 20;
-    static const size_t idx_bg_end   = SIR_NUM16_COLOR_MAPPINGS - 1;
-
-    size_t low  = 0;
-    size_t high = 0;
-
-    if ((style & _SIRS_ATTR_MASK) != 0) {
-        /* looking up an attribute */
-        low  = idx_attr_start;
-        high = idx_attr_end;
-    } else if ((style & _SIRS_FG_MASK) != 0) {
-        /* looking up a foreground color */
-        low  = idx_fg_start;
-        high = idx_fg_end;
-    } else if ((style & _SIRS_BG_MASK) != 0) {
-        /* looking up a background color */
-        low  = idx_bg_start;
-        high = idx_bg_end;
-    }
+bool _sir_formatstyle(sir_colormode mode, const sir_textstyle* style,
+    char buf[SIR_MAXSTYLE]) {
+    if (!_sir_validtextstyle(mode, style))
+        return false;
 
-    /* skip if not assigned due to invalid input. */
-    if (high > low) {
-        for (size_t n = low; n <= high; n++) {
-        if (sir_style_16color_map[n].from == style)
-            return sir_style_16color_map[n].to;
+    _sir_resetstr(buf);
+
+    bool formatted = false;
+    switch (mode) {
+        case SIRCM_16:
+            /* \x1b[attr;fg;bgm */
+            return 0 < snprintf(buf, SIR_MAXSTYLE, "%s%d;%d;%dm", SIR_ESC,
+                style->attr, _sir_mkansifgcolor(style->fg), _sir_mkansibgcolor(style->bg));
+        break;
+        case SIRCM_256: {
+            /* \x1b[attr;38;5;fg;48;5;bgm */
+            return 0 < snprintf(buf, SIR_MAXSTYLE, "%s%d;%d;5;%d;%d;5;%dm",
+                SIR_ESC, style->attr, _sir_getansifgcmd(style->fg), style->fg,
+                _sir_getansibgcmd(style->bg), style->bg);
         }
+        case SIRCM_RGB: {
```

Comment:
This is bothering me. that buffer is 39 bytes, and those are a lot of integers. I think instead of making it the number of bytes it should be in theory, it should be the maximum that could expand to if every integer were maxed out

Reply:
I don't trust snprintf to do the right thing. Best case scenario the log is jumbled up; worst–buffer overrun

Reply:
I think what I should do is make `get*fromcolor` explicitly return `uint8_t`, and use `PRId8` here, so I can rest easy